In [1]:
import os
from sqlalchemy import create_engine


user = os.getenv('MYSQL_USER')
password = os.getenv('MYSQL_PASS')
url = f'mysql+pymysql://{user}:{password}@localhost:3306/telecom'

engine = create_engine(url)
%load_ext sql
%sql engine
%config SqlMagic.autopandas = True

There's a new jupysql version available (0.10.3), you're running 0.9.1. To upgrade: pip install jupysql --upgrade


Found pyproject.toml from '/'

# Tower Information

In [3]:
%sql SHOW TABLES

Running query in 'mysql+pymysql://r12323011:***@localhost:3306/telecom'

12 rows affected.

Tables_in_telecom
0                     cinfo_xy
1          sample_sight_201308
2           sample_user_201308
3   serv_acct_item_0838_201307
4   serv_acct_item_0838_201308
5   serv_acct_item_0838_201406
6                  target_node
7      tb_asz_cdma_0838_201308
8      tb_asz_cdma_0838_201406
9                        tower
10                tower_center
11            tower_hot_201308

# Quick Look at various tables of tower

In [10]:
%sql  SELECT * FROM cinfo_xy LIMIT 10;

Running query in 'mysql+pymysql://r12323011:***@localhost:3306/telecom'

10 rows affected.

CELL_10 CELL_16      LON      LAT        X        Y
0       1    0001  103.993  30.7214  403.546  3400.52
1       2    0002  103.993  30.7214  403.546  3400.52
2       3    0003  103.993  30.7214  403.546  3400.52
3       4    0004  104.161  30.6291  419.562  3390.16
4       5    0005  104.161  30.6291  419.562  3390.16
5       6    0006  104.161  30.6291  419.562  3390.16
6       7    0007  103.681  30.9488  373.957  3426.05
7       8    0008  103.681  30.9488  373.957  3426.05
8       9    0009  103.681  30.9488  373.957  3426.05
9      10    000A  103.647  30.8490  370.574  3415.02

In [11]:
%sql  SELECT * FROM tower LIMIT 10;

Running query in 'mysql+pymysql://r12323011:***@localhost:3306/telecom'

10 rows affected.

CELL_10 CELL_16      LON      LAT        X        Y  DEYANG_CENTER
0       1    0001  103.993  30.7214  403.546  3400.52              0
1       2    0002  103.993  30.7214  403.546  3400.52              0
2       3    0003  103.993  30.7214  403.546  3400.52              0
3       4    0004  104.161  30.6291  419.562  3390.16              0
4       5    0005  104.161  30.6291  419.562  3390.16              0
5       6    0006  104.161  30.6291  419.562  3390.16              0
6       7    0007  103.681  30.9488  373.957  3426.05              0
7       8    0008  103.681  30.9488  373.957  3426.05              0
8       9    0009  103.681  30.9488  373.957  3426.05              0
9      10    000A  103.647  30.8490  370.574  3415.02              0

In [12]:
%sql  SELECT * FROM tower_center LIMIT 10;

Running query in 'mysql+pymysql://r12323011:***@localhost:3306/telecom'

10 rows affected.

CELL_10 CELL_16      LON      LAT        X        Y
0   13401    3459  104.374  31.1269  440.293  3445.21
1   13402    345A  104.374  31.1269  440.293  3445.21
2   13403    345B  104.374  31.1269  440.293  3445.21
3   13407    345F  104.415  31.1127  444.195  3443.62
4   13408    3460  104.415  31.1127  444.195  3443.62
5   13409    3461  104.415  31.1127  444.195  3443.62
6   13410    3462  104.389  31.1355  441.729  3446.16
7   13411    3463  104.389  31.1355  441.729  3446.16
8   13412    3464  104.389  31.1355  441.729  3446.16
9   13422    346E  104.382  31.0913  441.034  3441.26

In [13]:
%sql  SELECT * FROM tower_hot_201308 LIMIT 10;

Running query in 'mysql+pymysql://r12323011:***@localhost:3306/telecom'

10 rows affected.

CELL_10 CELL_16      LON      LAT        X        Y  MAIN_CENTER_FLAG  \
0       1    0001  103.993  30.7214  403.546  3400.52                 0   
1       2    0002  103.993  30.7214  403.546  3400.52                 0   
2       3    0003  103.993  30.7214  403.546  3400.52                 0   
3       4    0004  104.161  30.6291  419.562  3390.16                 0   
4       5    0005  104.161  30.6291  419.562  3390.16                 0   
5       6    0006  104.161  30.6291  419.562  3390.16                 0   
6       7    0007  103.681  30.9488  373.957  3426.05                 0   
7       8    0008  103.681  30.9488  373.957  3426.05                 0   
8       9    0009  103.681  30.9488  373.957  3426.05                 0   
9      10    000A  103.647  30.8490  370.574  3415.02                 0   

   HOT_FLAG  HOT_MAIN_FLAG  HOT_OTHER_FLAG  
0         0              0               0  
1         0              0               0  
2         0              0               0  
3         0              0               0  
4         0              0               0  
5         0              0               0  
6         0              0               0  
7         0              0               0  
8         0              0               0  
9         0              0               0

# number of observations

In [14]:
%%sql

SELECT 
    (SELECT COUNT(*) FROM cinfo_xy) AS cinfo_xy,
    (SELECT COUNT(*) FROM tower) AS tower,
    (SELECT COUNT(*) FROM tower_center) AS tower_center,
    (SELECT COUNT(*) FROM tower_hot_201308) AS tower_hot_201308;

Running query in 'mysql+pymysql://r12323011:***@localhost:3306/telecom'

1 rows affected.

cinfo_xy  tower  tower_center  tower_hot_201308
0     40569  40569           427             40569

we can see that the towers in the table *tower_center* those in *tower* table with center flag


In [15]:
%%sql
SELECT 
    COUNT(*)
FROM 
    tower
WHERE DEYANG_CENTER = 1;

Running query in 'mysql+pymysql://r12323011:***@localhost:3306/telecom'

1 rows affected.

COUNT(*)
0       427

# number of observations for each flag in table tower_hot_201308

In [16]:
%%sql

SELECT
    (SELECT COUNT(*) FROM tower_hot_201308 WHERE MAIN_CENTER_FLAG = 1) AS MAIN_CENTER_FLAG,
    (SELECT COUNT(*) FROM tower_hot_201308 WHERE HOT_FLAG = 1) AS HOT_FLAG,
    (SELECT COUNT(*) FROM tower_hot_201308 WHERE HOT_MAIN_FLAG = 1) AS HOT_MAIN_FLAG,
    (SELECT COUNT(*) FROM tower_hot_201308 WHERE HOT_OTHER_FLAG = 1) AS HOT_OTHER_FLAG;

Running query in 'mysql+pymysql://r12323011:***@localhost:3306/telecom'

1 rows affected.

MAIN_CENTER_FLAG  HOT_FLAG  HOT_MAIN_FLAG  HOT_OTHER_FLAG
0               427       425            132             293

- 425 = 132 + 293
- which means the most of the heavy traffic tower is not in the center of deyang
- notice that the hot flag is calucated from 10 a.m. to 10 p.m.

In [44]:
group1 = cudf.merge(group1, meta_tower, on='cell_id')
group1

serv_id client_nbr  time  day_of_week calling_nbr  \
0        9744kdpd   5qz4l0kf    17            5    ju3titmy   
1        98ep43zd   6q1dk05c    17            5    irrf79qp   
2        a5lx9g9j   4gogl0n7    17            5    96vlehrp   
3        aa08jix0   h1sz3a80    17            5    fyv5krgs   
4        9nj82tj4   ltms5rt3    17            5    eweobjtv   
...           ...        ...   ...          ...         ...   
3781632  90gkc8qc   ence3aj3    20            7    3camjtti   
3781633  ad33g38g   nfk232w7    20            7    gw3a8i6o   
3781634  aa01jy76   jotskt5y    20            7    8jbbfv5g   
3781635  8sd06h0i   9k1v32yt    20            7    b96ok48m   
3781636  8ksv3gdc   c2r23aga    19            7    h83khwzx   

         calling_area_code  duration  called_area_code called_nbr cell_id  \
0                      871        63               838   5qz4l0kf    3516   
1                      827        35               838   6q1dk05c    0122   
2                       28        33               838   4gogl0n7    3CAE   
3                       28        70               838   h1sz3a80    01E0   
4                       28        16               838   ltms5rt3    07C0   
...                    ...       ...               ...        ...     ...   
3781632                818       103               838   ence3aj3    355D   
3781633                 28        26               838   nfk232w7    3530   
3781634                894       900               838   jotskt5y    35A7   
3781635                 28        89               838   9k1v32yt    37F6   
3781636                 28      1692               838   c2r23aga    3B29   

         CELL_10 CELL_16      LON      LAT        X        Y  DEYANG_CENTER  \
0          13590    3516  104.287  30.9648  431.880  3427.29              0   
1            290    0122  104.413  30.8044  443.824  3409.44              0   
2          15534    3CAE  104.168  30.9560  420.503  3426.39              0   
3            480    01E0  104.261  30.8860  429.338  3418.57              0   
4           1984    07C0  104.119  30.6705  415.571  3394.78              0   
...          ...     ...      ...      ...      ...      ...            ...   
3781632    13661    355D  104.273  31.1376  430.667  3446.46              0   
3781633    13616    3530  104.298  31.0488  432.989  3436.60              0   
3781634    13735    35A7  104.674  31.0251  468.874  3433.80              0   
3781635    14326    37F6  104.171  31.1278  420.931  3445.44              0   
3781636    15145    3B29  104.541  31.0170  456.171  3432.95              0   

         call_area_code  
0                   838  
1                    28  
2                   838  
3                    28  
4                    28  
...                 ...  
3781632             838  
3781633             838  
3781634             838  
3781635             838  
3781636             838  

[3781637 rows x 18 columns]

In [48]:
len(group1[group1['calling_area_code'] == group1['call_area_code']])

715058

In [50]:
e = len(group1[group1['calling_area_code'] == group1['call_area_code']]) /3781637 * 100
e

18.90868954370819

In [51]:
 len(group1[group1['called_area_code'] == group1['call_area_code']])

2768451

In [53]:
f = len(group1[group1['called_area_code'] == group1['call_area_code']]) / 3781637 * 100
f

73.20774045737335

In [54]:
100 - e - f

7.88356999891846

In [55]:
len(group1[group1['client_nbr'] == group1['called_nbr']])

3781637

# inspect the locatiion(lat, lon position) distibution of meta tower

In [1]:
import os
os.chdir('../../../')

In [3]:
import cudf

In [4]:
meta_tower = cudf.read_csv('data/processed/meta_tower.csv')

meta_tower

CELL_10 CELL_16      LON      LAT        X        Y  DEYANG_CENTER  \
0            1    0001  103.993  30.7214  403.546  3400.52              0   
1            2    0002  103.993  30.7214  403.546  3400.52              0   
2            3    0003  103.993  30.7214  403.546  3400.52              0   
3            4    0004  104.161  30.6291  419.562  3390.16              0   
4            5    0005  104.161  30.6291  419.562  3390.16              0   
...        ...     ...      ...      ...      ...      ...            ...   
40564    65512    FFE8  103.966  30.5730  400.809  3384.09              0   
40565    65520    FFF0  103.980  30.6016  402.180  3387.25              0   
40566    65522    FFF2  103.966  30.5730  400.809  3384.09              0   
40567    65524    FFF4  104.066  30.5781  410.407  3384.57              0   
40568    65527    FFF7  104.060  30.6573  409.905  3393.36              0   

       call_area_code  
0                  28  
1                  28  
2                  28  
3                  28  
4                  28  
...               ...  
40564              28  
40565              28  
40566              28  
40567              28  
40568              28  

[40569 rows x 8 columns]

In [5]:
meta_tower['call_area_code'].nunique()

25

In [9]:
meta_tower.groupby('call_area_code').size().sort_values(ascending=False)

call_area_code
28     11338
830     2433
831     2370
817     2317
833     2289
838     2195
818     1862
839     1672
832     1561
834     1547
827     1521
813     1481
816     1363
826     1292
835     1284
812     1117
825     1093
837      910
836      888
23        22
878        6
870        3
852        2
939        2
916        1
dtype: int32